rfcs/0004-insecure-http/README.md

```diff
@@ -0,0 +1,173 @@
+This shall be enabled by adding a new flag to all controllers `--allow-insecure-http`.
+The default value of this flag shall be `false`. This would ensure that there is no breaking change with controllers 
+still being able to access non-TLS endpoints. To disable this behavior and enforce the use of HTTPS connections, users would
+have to explicitly pass the flag to the controller:
+
+```yaml
+spec:
+  template:
+    spec:
+      containers:
+      - name: manager
+        image: fluxcd/source-controller
+        args:
+          - --watch-all-namespaces
+          - --log-level=info
+          - --log-encoding=json
+          - --enable-leader-election
+          - --storage-path=/data
+          - --storage-adv-addr=source-controller.$(RUNTIME_NAMESPACE).svc.cluster.local.
+          - --allow-insecure-http=false
+```
+
+### Objects
+Some Flux objects, like `GitRepository`, provide a field for specifying a URL, and the URL would contain the scheme.
+In such cases, the scheme can be used for inferring the transport type of the connection and consequently,
+whether to use HTTP or HTTPS connections for that object.
+But there are a few objects that don't allow such behavior, for example:
+
+* `ImageRepository`: It provides a field, `.spec.image`, which is used for specifying the URL of the image present on
+a container registry. But any URL containing a scheme is considered invalid and HTTPS is the default transport used.
+This prevents users from using images present on insecure registries.
+* OCI `HelmRepository`: When using an OCI registry as a Helm repository, the `.spec.url` is expected to begin with `oci://`.
+Since the scheme part of the URL is used to specify the type of `HelmRepository`, there is no way for users to specify
+that the registry is hosted at a non-TLS endpoint.
+
+For such objects, we shall introduce a new boolean field `.spec.insecure`, which shall be `false` by default. Users that
+need their object to point to an HTTP endpoint, can set this to `true`.
+
+### CLI
+The Flux CLI offers several commands for creating Flux specific resources. Some of these commands may involve specifying
+an endpoint such as creating an `OCIRepository`:
+
+```sh
+ flux create source oci podinfo \
+    --url=oci://ghcr.io/stefanprodan/manifests/podinfo \
+    --tag=6.1.6 \
+    --interval=10m
+```
+
+Since these commands essentially create object definitions, the CLI should offer a boolean flag `--insecure`
+for relevant objects, which will be used for specifying the value of `.spec.insecure` of such objects.
+
+### Precedence & Validitiy
+Objects with `.spec.insecure` as `true `will only be allowed if HTTP connections are allowed at the controller
+level.
+Similarly, an object can have`.spec.insecure` as `true` only if the Saas/Cloud provider allows HTTP connections.
+For example using a `Bucket` with its `.spec.provider` set to `azure` would be invalid, since Azure doesn't allow
+HTTP connections.
```

Comment:
The Azure SDK supports HTTP access:
https://github.com/Azure/azure-sdk-for-go/blob/9b91fa8fd891a3364779d8d32f2a13dfa98dc6f4/storage/client.go#L269

But currently we dont use the `.Insecure` field for anything related to that specific case. As part of this RFC we could establish errors that should be observed on such situations.

Reply:
The Azure SDK referenced by you here is deprecated: https://github.com/Azure/azure-sdk-for-go/tree/9b91fa8fd891a3364779d8d32f2a13dfa98dc6f4/storage
We currently use the recommended sdk: https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/storage/azblob, which does not support HTTP connections to my knowledege.